File: src/main/java/org/fountainmc/api/entity/data/EntityDataFactory.java
package org.fountainmc.api.entity.data;

import java.util.List;
import java.util.UUID;
import javax.annotation.concurrent.ThreadSafe;

import org.fountainmc.api.GameMode;
import org.fountainmc.api.NonnullByDefault;
import org.fountainmc.api.Server;
import org.fountainmc.api.entity.EntityType;

@ThreadSafe
@NonnullByDefault
public interface EntityDataFactory {

    Server getServer();

    EntityData createEntityData(
            EntityType<?> type,
            float pitch,
            float yaw,
            List<? extends EntityData> passengers,
            int ticksOnFire,
            boolean immuneToFire
    );

    LivingEntityData createLivingEntityData(
            EntityData entityData,
            double health,
            double maxHealth
    );

    PlayerData createPlayerData(
            LivingEntityData entityData,
            String name,
            UUID uniqueId,
            GameMode gameMode,
            boolean canFly,
            float percentageToNextExperienceLevel,
            int experienceLeve
    );


}

File: src/main/kotlin/org/fountainmc/api/entity/data/EntityDataFactory.kt
package org.fountainmc.api.entity.data

import java.util.UUID
import javax.annotation.concurrent.ThreadSafe

import org.fountainmc.api.GameMode
import org.fountainmc.api.Server
import org.fountainmc.api.entity.EntityType
import org.fountainmc.api.entity.LivingEntity

// TODO: Consider builder pattern?
@ThreadSafe
interface EntityDataFactory {

    val server: Server

    fun createEntityData(
            type: EntityType<*>,
            pitch: Float = 0.0f,
            yaw: Float = 0.0f,
            passengers: List<EntityData> = emptyList(),
            ticksOnFire: Int = 0,
            immuneToFire: Boolean = false
    ): EntityData

    fun createLivingEntityData(
            type: EntityType<LivingEntity>,
            entityData: EntityData = createEntityData(type),
            health: Double = maxHealth,
            maxHealth: Double = entityData.entityType.maxHealth
    ): LivingEntityData

    fun createPlayerData(
            entityData: LivingEntityData = createLivingEntityData(type = EntityType.PLAYER),
            name: String,
            uniqueId: UUID,
            gameMode: GameMode = GameMode.SURVIVAL,
            canFly: Boolean = false,
            percentageToNextExperienceLevel: Float = 0.0f,
            experienceLevel: Int = 0
    ): PlayerData
}
